cuda_phase-retrieval_class/src/phase_algo.hpp

```diff
@@ -0,0 +1,633 @@
+
+        double y = get_real(normalize(random_phase[idx], dimension)); //Get real version of normalized complex number
+
+        /**
+       * \b Process
+       * 1. Get previous image based on current iteration
+       */
+        if(iter == 0) image_output_temp = y;
+        else image_output_temp = image_output[idx];
+
+        /**
+        * 2. Updates for elements that satisfy object domain constraint
+        */
+        if(phase_mode == OutputOutput || phase_mode == Hybrid) image_output[idx] = y;
+
+        /**
+        * 3. Find elements that violate object domain constraints or are not masked 
+        */
+        if(mask[idx] <= 0) logical_not_mask = true;
+        else if(mask[idx] >= 1) logical_not_mask = false;
+
+        /**
+        * 4. Check if any element y is less than zero 
+        */
+        if(y < 0) y_less_than_zero = true;
+        else if(y >= 0) y_less_than_zero = false;
+
+        /**
+        * 5. Use "and" logical to check the "less than zero y" and the mask
+        */   
+        if(y_less_than_zero == true && mask[idx] >= 1) logical_and = true;
+        else logical_and = false;
+
+        /**
+        * 6. Determine indices value
+        */  
+        if(logical_and == false && logical_not_mask == false) indices = false;
+        else indices = true;
+
+        /**
+        * 7. Updates for elements that violate object domain constraints
+        */ 
+        if(indices == true)
+        {
+            if(phase_mode == Hybrid || phase_mode == InputOutput)
+            {
+                image_output[idx] = image_output_temp-beta*y;
+            }
+            else if(phase_mode == OutputOutput)
+            {
+                image_output[idx] = y-beta*y;
+            }
+        }
+
+        random_phase[idx] = get_complex(image_output[idx]);
+    }
+}
+
+/**
+* \brief CUFFT error checking
+* \param cufft_process Result of a CUFFT operation
+*/ 
+void CUFFT_CHECK(cufftResult cufft_process)
+{
+    if(cufft_process != CUFFT_SUCCESS) std::cout<<cufft_process<<std::endl;
+}
+
+/**
+* \brief Set number of Streaming Multiprocessor for CUDA Kernel
+*/ 
+int get_number_of_cuda_sm()
+{
+    int devId, numSMs;
+    cudaGetDevice(&devId);
+    cudaDeviceGetAttribute( &numSMs, cudaDevAttrMultiProcessorCount, devId);
+    return numSMs;
+}
+
+/**
+* \brief Reinterpret a complex pointer from standard complex to CUDA FFT
+* \param ptr a standard complex number
+* \return CUDA FFT version of the standard complex number
+*/ 
+template<typename TInputData, typename TOutputData>
+TOutputData * convertToCUFFT(TInputData * ptr){}
+
+template<>
+cufftDoubleComplex *convertToCUFFT(std::complex<double> * ptr)
+{  
+    return reinterpret_cast<cufftDoubleComplex *>(ptr);
+}
+
+/**
+* \brief Generate a cupy array with 0 as the values
+* \param shape vector representing dimension and size
+* \return python object, which is a cupy array
+*/
+template<typename T> 
+pybind11::object cupy_allocate(std::vector<int> shape)
```

Comment:
This function should be moved to an extra header file in the folder `include`

Reply:
please put this function in a new, extra header `cupy_allocate.hpp`

Reply:
please check this. cupy_allocate is in includes